presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,588 @@
+        Set<Symbol> childSourceSymbols = child.getSources().stream().map(PlanNode::getOutputSymbols).flatMap(Collection::stream).collect(toImmutableSet());
+
+        return allDereferencesInProject.entrySet().stream()
+                .filter(entry -> childSourceSymbols.contains(getBase(entry.getKey())))
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private static PlanNode createProjectBelow(PlanNode planNode, Assignments dereferences, PlanNodeIdAllocator idAllocator)
+    {
+        if (dereferences.isEmpty()) {
+            return planNode;
+        }
+        return new ProjectNode(idAllocator.getNextId(), planNode, Assignments.builder().putIdentities(planNode.getOutputSymbols()).putAll(dereferences).build());
+    }
+
+    private static class DereferenceReplacer
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<DereferenceExpression, Symbol> expressions;
+
+        DereferenceReplacer(Map<DereferenceExpression, Symbol> expressions)
+        {
+            this.expressions = requireNonNull(expressions, "expressions is null");
+        }
+
+        @Override
+        public Expression rewriteDereferenceExpression(DereferenceExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (expressions.containsKey(node)) {
+                return expressions.get(node).toSymbolReference();
+            }
+            return treeRewriter.defaultRewrite(node, context);
+        }
+    }
+
+    private static List<DereferenceExpression> extractDereferenceExpressions(Expression expression)
+    {
+        ImmutableList.Builder<DereferenceExpression> builder = ImmutableList.builder();
+        new DefaultExpressionTraversalVisitor<Void, ImmutableList.Builder<DereferenceExpression>>()
+        {
+            @Override
+            protected Void visitDereferenceExpression(DereferenceExpression node, ImmutableList.Builder<DereferenceExpression> context)
+            {
+                context.add(node);
+                return null;
+            }
+        }.process(expression, builder);
+        return builder.build();
+    }
+
+    private static Map<DereferenceExpression, Symbol> getDereferenceSymbolMap(Collection<Expression> expressions, Context context, TypeAnalyzer typeAnalyzer)
+    {
+        Set<DereferenceExpression> dereferences = expressions.stream()
+                .flatMap(expression -> extractDereferenceExpressions(expression).stream())
+                .filter(PushDownDereferences::validPushDown)
+                .collect(toImmutableSet());
+
+        // When nested child and parent dereferences both exist, Pushdown rule will be trigger one more time
+        // and lead to runtime error. E.g. [msg.foo, msg.foo.bar] => [exp, exp.bar] (should stop here but
+        // since there are still dereferences, pushdown rule will trigger again)
+        if (dereferences.stream().anyMatch(exp -> baseExists(exp, dereferences))) {
```

Comment:
Yes. it will not pushdown `a.b`.  I don't know how to stop iteration in this case so this is just a walkaround, not final solution.

> We can return ImmutableMap.of() in cases where all bases of all dereference expressions are symbolreferences
How does this work? It will stop the first time for `[f(x.y) + a.b]` because `x` and `a` are both symbol references.

Reply:
I think optimizing such pushdowns here will impact the effectiveness of all individual pushdown rules, since this method is being used everywhere.